sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/RecoverableReactorConnection.java

```diff
@@ -0,0 +1,249 @@
+                                        String entityPath,
+                                        AmqpRetryPolicy retryPolicy,
+                                        Map<String, Object> loggingContext) {
+        this.fullyQualifiedNamespace = Objects.requireNonNull(fullyQualifiedNamespace, "'fullyQualifiedNamespace' cannot be null.");
+        // Note: FQDN, (to an extent) entity-path are generic enough, but if we find more connection description
+        // parameters that are non-generic, i.e., specific to individual messaging services, then consider creating
+        // dedicated POJO types to pass around connection description parameters in corresponding libraries rather
+        // than polluting shared 'RecoverableReactorConnection' type.
+        this.entityPath = entityPath;
+        Objects.requireNonNull(retryPolicy, "'retryPolicy' cannot be null.");
+        this.retryOptions = retryPolicy.getRetryOptions();
+        this.errorContext = new AmqpErrorContext(fullyQualifiedNamespace);
+        this.logger = new ClientLogger(getClass(), Objects.requireNonNull(loggingContext, "'loggingContext' cannot be null."));
+        Objects.requireNonNull(connectionSupplier, "'connectionSupplier' cannot be null.");
+        final Mono<C> newConnection = Mono.fromSupplier(() -> {
+            if (terminated) {
+                logger.info("Connection recovery support is terminated, dropping the request for new connection.");
+                throw TERMINATED_ERROR;
+            } else {
+                return connectionSupplier.get();
+            }
+        });
+
+        this.createOrGetCachedConnection = newConnection
+            .flatMap(c -> {
+                logger.atInfo()
+                    .addKeyValue(CONNECTION_ID_KEY, c.getId())
+                    .log("Waiting to connect and become active.");
+
+                return c.connectAndAwaitToActive()
+                    .doOnCancel(() -> {
+                        if (!c.isDisposed()) {
+                            c.closeAsync(closeSignal("The connection request was canceled while waiting to active."))
+                                .subscribe();
+                        }
+                    });
+            })
+            .retryWhen(retryWhenSpec(retryPolicy))
+            .<C>handle((c, sink) -> {
+                @SuppressWarnings("unchecked")
+                final C connection = (C) c;
+                currentConnection = connection;
+                if (terminated) {
+                    connection.closeAsync(closeSignal("Connection recovery support is terminated.")).subscribe();
+                    sink.error(TERMINATED_ERROR);
+                } else {
+                    logger.atInfo()
+                        .addKeyValue(CONNECTION_ID_KEY, c.getId())
+                        .log("Emitting the new active connection.");
+                    sink.next(connection);
+                }
+            }).cacheInvalidateIf(c -> {
```

Comment:
Finally! I wish this operator existed a long time ago... would have simplified writing this.

Reply:
Yea, I recently came across this operator while exploring the other reactor buffering reduction work. The operator is implemented to simplify use case (e.g., in RabbitMQ, RSocket) similar to ours https://github.com/reactor/reactor-core/issues/2324

Reply:
I've updated the pr description so others can understand the reason behind original ACP implementation.